Add ability to configure kubeadm's ignored pre-flight errors via InitConfiguration and JoinConfiguration

marccarre opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature
/sig cluster-lifecycle

**What this PR does / why we need it**:
This adds a more declarative way to configure ignored pre-flight errors -- via `InitConfiguration` and `JoinConfiguration`, in addition to the "legacy" CLI flag -- and therefore simplifies programmatic usage of `kubeadm`: just one configuration file to generate & to pass to `kubeadm`.

**Which issue(s) this PR fixes**:
Fixes #74246.
Fixes kubernetes/kubeadm/issues/1460.

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm's ignored pre-flight errors can now be configured via InitConfiguration and JoinConfiguration.
```

marccarre committed:
kubeadm errors now ignorable via v1beta2 config files

Specifically, IgnorePreflightErrors in {Init,Join}Configuration's NodeRegistrationOptions can be used to achieve this.
See also: https://docs.google.com/document/d/1XnP67oO1i9VcDIpw42IzptnJsc5OQM-HTf8cVcjCR2w/edit

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go:
```diff
@@ -34,6 +34,9 @@ type InitConfiguration struct {
 	ClusterConfiguration `json:"-"`
 
+	// IgnorePreflightErrors holds the list of pre-flight errors to ignore during kubeadm's initialisation.
+	IgnorePreflightErrors []string `json:"ignorePreflightErrors,omitempty"`
```
this is already planned in the v1beta2 config, see:
https://github.com/kubernetes/kubeadm/issues/1439

so technically we can accept this PR but only once the v1beta1 types are already created, and we cannot add this to v1beta1. today we are going to have a kubeadm 1.15 planning session at 4PM UTC.
https://github.com/kubernetes/community/tree/master/sig-cluster-lifecycle

marccarre replied:
TIL, thanks for pointing to this issue.
I have undone the changes on v1beta1, but now see the below error:
```console
$ make WHAT=cmd/kubeadm
[...]
E0321 15:15:07.858860   27590 conversion.go:667] Warning: could not find nor generate a final Conversion function for k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm.NodeRegistrationOptions -> k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha3.NodeRegistrationOptions
E0321 15:15:07.858900   27590 conversion.go:668]   the following fields need manual conversion:
E0321 15:15:07.858910   27590 conversion.go:670]       - IgnorePreflightErrors
E0321 15:15:07.972950   27590 conversion.go:667] Warning: could not find nor generate a final Conversion function for k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm.NodeRegistrationOptions -> k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1.NodeRegistrationOptions
E0321 15:15:07.972989   27590 conversion.go:668]   the following fields need manual conversion:
E0321 15:15:07.972996   27590 conversion.go:670]       - IgnorePreflightErrors
[...]
make: *** [all] Error 1
```
...and it isn't obvious where this manual conversion should be added. 
Any pointer?

marccarre replied:
Resolved now that I've rebased on top of latest `master`.

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -396,6 +399,14 @@ func newInitData(cmd *cobra.Command, args []string, options *initOptions, out io
 	}, nil
 }
 
+func union(lists ...[]string) []string {
```
i think we had a similar utility function in another location.

marccarre replied:
I'd be interested in it if you find it again. `git grep`-ing for `func Union`/`func union` didn't yield anything.

marccarre replied:
Alternatively, I could move this to some utility file. Any preference on the location? (I was thinking of moving it under `cmd/kubeadm/app/util/sets.go`)

kad replied:
you can do in the code above something like this:
```go

ignorePreflightErrors := sets.NewString(options.ignorePreflightErrors).Union(sets.NewString(cfg.NodeRegistration.IgnorePreflightErrors)).List()
```

marccarre replied:
Yes, I thought of this too, but found it relatively verbose, so thought I'd hide the details behind a function. I can do this if there is a strong opinion though. Thanks @kad for pointing this out in any case 👍

marccarre replied:
Resolved now that the union is done directly by `ValidateIgnorePreflightErrors` as suggested in https://github.com/kubernetes/kubernetes/pull/75499#issuecomment-492216224.

neolit123 reviewed (commented):
thanks for the PR @marccarre 
i think, this is pretty much what we want to do, except that it has to happen in v1beta2.

/assign @rosti @fabriziopandini 
/hold

neolit123 commented:
@marccarre 
please squash the commits into one.
/priority important-longterm

rosti reviewed (commented):
Thanks for your contribution @marccarre !
The thing is that we should not modify the existing v1beta1 config. We have to do that modification in the new v1beta2 config if we reach a consensus on that.
Also, my proposition is to have that field in the `NodeRegistrationOptions` only.

Let's hold this for now, until the time comes for it.
/hold

marccarre commented:
@rosti, @neolit123, thanks for your feedback ✨ 
I've addressed some of it (and will squash at the very end once everyone is happy about this PR), but now face build issues as per https://github.com/kubernetes/kubernetes/pull/75499#discussion_r267634175. Apologies for being new to `kubeadm`/`kubernetes`' development, but any pointer on how to resolve this would be very much appreciated.

rosti reviewed (commented):
Thanks for updating this @marccarre !
However, we need to hold on until the new config (probably v1beta2) becomes available. More developments on this question are to come probably next week.
We need to add the new field in that config and the internal one (not in v1beta1 or v1alpha3).
I'll ping you for a rebase and more guidance once we are ready for this.
It's just that we need to do a few more things before we can proceed with this.

marccarre commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -225,6 +225,9 @@ type NodeRegistrationOptions struct {
 	KubeletExtraArgs map[string]string
+
+	// IgnorePreflightErrors provides a slice of pre-flight errors to be ignored when the current node is registered.
+	IgnorePreflightErrors []string `json:"ignorePreflightErrors,omitempty"`
```
Ha! I had removed them initially when experimenting, but eventually added them back as I thought they didn't hurt. I've just removed them again. Thanks again for the feedback!

rosti replied:
The tag will be needed in the versioned config (v1beta2) though. But since that is non-existent as of today, there is not much you can do ATM.

marccarre replied:
Resolved: I had removed the tag on `cmd/kubeadm/app/apis/kubeadm/types.go` a while ago, and I've now added the field _with_ the tag in the `v1beta2` file.

marccarre commented:
> we need to hold on until the new config (probably v1beta2) becomes available

Sure, no worries. I'll wait then 🙂 

I haven't dug into the code generation logic (yet), but I'm guessing some of the build issues I'm seeing will change/disappear once `v1beta2` is added.

rosti commented:
> I haven't dug into the code generation logic (yet), but I'm guessing some of the build issues I'm seeing will change/disappear once `v1beta2` is added.

Mostly yes.

rosti commented:
@marccarre, with the merger of #76710 you are now able to continue working on this based on v1beta2.
In addition to that we need to verify the new field and make sure it cannot be set to "all" and therefore ignore all pre-flight errors from the config easily.
Thanks for taking this on!

kad commented:
I thought about this PR, and I think the better way to implement that will be to modify `validation.ValidateIgnorePreflightErrors` such, that it will take two arrays of strings as arguments. One from command line, one from config objects. Then, merging into one set can be done inside `validation.ValidateIgnorePreflightErrors` by iterating both of the arrays, practically getting rid of helper function `union`. Also it would be good to validate that array from config object will not contain 'all', as config object should not be used as "hide everything" workarround, instead of explicitly specifying what should be ignored.

marccarre commented:
@rosti, @fabriziopandini, @neolit123, @kad, I've just rebased this PR on top of latest `master`, added the necessary changes for `v1beta2` and addressed the feedback given so far (thank you all! 🙇). PTAL 🙂

kad commented:
Please squash commits. One for implementation/code changes part, one commit for auto-generated stuff.

rosti commented on cmd/kubeadm/app/apis/kubeadm/v1beta1/conversion.go:
```diff
@@ -45,3 +45,15 @@ func Convert_kubeadm_JoinControlPlane_To_v1beta1_JoinControlPlane(in *kubeadm.Jo
 	return nil
 }
+
+func Convert_kubeadm_NodeRegistrationOptions_To_v1beta1_NodeRegistrationOptions(in *kubeadm.NodeRegistrationOptions, out *NodeRegistrationOptions, s conversion.Scope) error {
```
This needs a fairly simple test case.

marccarre replied:
Done. Will push soon.

rosti commented on cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go:
```diff
@@ -215,6 +215,9 @@ type NodeRegistrationOptions struct {
 	// Flags have higher priority when parsing. These values are local and specific to the node kubeadm is executing on.
 	KubeletExtraArgs map[string]string `json:"kubeletExtraArgs,omitempty"`
+
+	// IgnorePreflightErrors provides a slice of pre-flight errors to be ignored when the current node is registered.
```
It's worth noting here that, although "all" is allowed on the command line, it's not allowed in the config.

marccarre replied:
Thanks for clarifying this. I indeed did not know this requirement.

marccarre replied:
Solved, I think. Will push soon.

rosti commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -464,12 +464,16 @@ func ValidateAPIEndpoint(c *kubeadm.APIEndpoint, fldPath *field.Path) field.Erro
-func ValidateIgnorePreflightErrors(ignorePreflightErrors []string) (sets.String, error) {
+// Multiple arrays of ignored pre-flight errors can be provided, as there may
+// be various sources for these, e.g. CLI options, configuration files, etc.
+func ValidateIgnorePreflightErrors(ignorePreflightErrorsArrays ...[]string) (sets.String, error) {
```
Ok, this is a little bit different, from what I expected.
The thing is that we want "all" to be an invalid value for errors specified in the config YAML.
Hence, it's beset to keep more or less the previous variant of the func, add a bool parameter to it (to tell if the error list is from the config file or the command line) and then denying "all" if the bool says it's from the config file.

kad replied:
@rosti, my suggestion was instead of `...[]string` to have two explicit arguments - one from command line, one from config. then to have two for loops, one processing config, then checking if 'all' is present in set (error if yes), then add overrides from command line. No need for bool here.

marccarre replied:
Sure, I'll go for this then.

marccarre replied:
Note that if we go for two explicit arguments (1), whether at this level, or in a wrapper function, maintainers would have to ensure arguments are passed in the right order.

If we go for a boolean (2), is may be a tiny bit more explicit, but the boolean would still be rather obscure, i.e. one would have to look at its name in the implementation to know its meaning.

A third approach (3), more verbose, but more explicit, would be to have two methods, e.g.`ValidateIgnorePreflightErrorsFromCLI` and `ValidateIgnorePreflightErrorsFromConfigFile` (or so, better names always welcome).

@kad, @rosti, please let me know your preferences.
(At the moment, I'm hesitating between (1) and (3).)

marccarre replied:
(1) currently implemented.

rosti commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -316,6 +311,13 @@ func newInitData(cmd *cobra.Command, args []string, options *initOptions, out io
 		return nil, err
 	}
 
+	ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(options.ignorePreflightErrors, cfg.NodeRegistration.IgnorePreflightErrors)
```
Ideally we should call this twice - once for command line supplied errors and once for config file supplied ones. This is because of the "all" keyword (it must not be specified in the config file).
After that, we should have a couple of sets, that need merging. If the command line set has "all" in it, the resulting set should be consisting only of `{"all"}`. Otherwise it's a simple merge of both. Probably a helper func should do the trick.

kad replied:
@rosti  if config has something specific, and command line says `ignore all` we still need to fail. We shouldn't allow to mix 'all' with anything else. Let admin think twice (or more times) when admin tries to mask all errors.

marccarre replied:
This indeed sounds like a desirable behaviour to me.

rosti commented on cmd/kubeadm/app/cmd/init_test.go:
```diff
@@ -157,3 +185,14 @@ func TestNewInitData(t *testing.T) {
 	}
 }
+
+func expectedInitIgnorePreflightErrors(expected sets.String) func(t *testing.T, data *initData) {
```
You can have the `expected` parameter as a string slice and then convert it to a set with `NewString`. This will make the calls less verbose by omitting the extra `NewString` call in them.

marccarre replied:
Indeed better, well spotted 👍 
Will push soonish.

rosti reviewed (commented):
Thanks for the update @marccarre !
So far it looks great!
However, we need to deny the ability to specify "all" in the config file (but still permit it in the command line). Once that is done we can merge it.

rosti commented:
/ok-to-test

rosti commented:
> I thought about this PR, and I think the better way to implement that will be to modify `validation.ValidateIgnorePreflightErrors` such, that it will take two arrays of strings as arguments. One from command line, one from config objects. Then, merging into one set can be done inside `validation.ValidateIgnorePreflightErrors` by iterating both of the arrays, practically getting rid of helper function `union`. Also it would be good to validate that array from config object will not contain 'all', as config object should not be used as "hide everything" workarround, instead of explicitly specifying what should be ignored.

That's one way of doing it. 
I would like to keep things simple and scope of functionality clear, hence `ValidateIgnorePreflightErrors` should still perform only validation of errors, but with a bool param stating the origin of the ignored errors. This is where "all" from a config file should be checked.
Then a wrapper func, that takes a couple of sets (one from command line, another from the config) would call `ValidateIgnorePreflightErrors` twice. If the "all" keyword is present in the command line errors, it's returned as a set with a single element. Otherwise, the sets need to be combined.

kad commented:
> Then a wrapper func, that takes a couple of sets (one from command line, another from the config) would call `ValidateIgnorePreflightErrors` twice. If the "all" keyword is present in the command line errors, it's returned as a set with a single element. Otherwise, the sets need to be combined.

This will require third iteration over the combined set with check, for mixture or `all` plus individual check, like it is currently performed in `validation.ValidateIgnorePreflightErrors`.

marccarre committed:
kubeadm errors now ignorable via v1beta2 config files (generated diff)

This commit contains only changes generated by the build process.
Nothing here was manually changed.

Changes made to:
```
cmd/kubeadm/app/apis/kubeadm/validation/BUILD
cmd/kubeadm/app/cmd/BUILD
```
were generated by running:
````
./hack/update-bazel.sh
```

marccarre commented:
I hope to have addressed both your requirements.
PTAL, @rosti, @kad, and thanks again for your feedback! ✨ 
(Previous push had some tests failing, if they still are after this push, I'll then look into this.)

kad commented on cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go:
```diff
@@ -702,6 +703,100 @@ func TestValidateFeatureGates(t *testing.T) {
+		{ // skip all checks
+			[]string{"all"},
+			[]string{"a", "b"},
+			sets.NewString("all"),
```
not correct expected result. should be error. 'all' should not be mixed with anything else.

marccarre replied:
My understanding was that **if** `all` is provided via the CLI, it should override the config file.
Is that not what we'd like?

kad replied:
This forces to limit usage of 'all' as much as possible. Easier to add any individual check, rather than trying to blindly ignoring all possible errors from checks without thinking. We seen a lot of 'guides' in the internet where they all started with 'kubeadm .... --skip-preflight-checks' and then trying to deal with problems that were actually reported by preflight checks as errors/warnings.
So, morale for admins: if you really know what you're doing, think again, and if still think that error must be ignored - add individual error to the list. If you copy&pasting from somewhere 'all', think 3 times, are you really want to do that.

kad commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -463,8 +463,30 @@ func ValidateAPIEndpoint(c *kubeadm.APIEndpoint, fldPath *field.Path) field.Erro
+		// By construction in validateIgnorePreflightErrorsFromCLI, if "all" is
+		// in the set from CLI, then it is the ONLY value it contains, hence we
+		// return ONLY this set.
+		return ignoreErrorsFromCLI, nil
```
If 'all' is set in CLI and config has something else in addition, it is an error situation. 
```go
func ValidateIgnorePreflightErrors(ignorePreflightErrorsFromCLI, ignorePreflightErrorsFromConfigFile []string) (sets.String, error) {
	ignoreErrors := sets.NewString()
	allErrs := field.ErrorList{}

	for _, item := range ignorePreflightErrorsFromConfigFile {
		ignoreErrors.Insert(strings.ToLower(item)) // parameters are case insensitive
	}

	if ignoreErrors.Has("all") {
		// "all" is forbidden in config files. Administrators should use an
		// explicit list of errors they want to ignore, as it can be risky to
		// mask all errors in such a way. Hence, we return an error:
		allErrs = append(allErrs, field.Invalid(field.NewPath("ignorePreflightErrors"), "all", "'all' cannot be used in configuration file"))
        }

	for _, item = range ignorePreflightErrorsFromCLI {
		ignoreErrors.Insert(strings.ToLower(item)) // parameters are case insensitive
	}

	if ignoreErrors.Has("all") && ignoreErrors.Len() > 1 {
		allErrs = append(allErrs, field.Invalid(field.NewPath("ignore-preflight-errors"), strings.Join(ignoreErrors.List(), ","), "don't specify individual checks if 'all' is used"))
	}

	return ignoreErrors, allErrs.ToAggregate()
}
```

kad commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -463,20 +463,35 @@ func ValidateAPIEndpoint(c *kubeadm.APIEndpoint, fldPath *field.Path) field.Erro
 	}
 
-	return ignoreErrors, allErrs.ToAggregate()
+	return ignoreErrors, nil
```
Small nit, not critical, but if you'll be changing due to failed tests, better to have only one `return` statement, as in previous version of function, and return all possible errors at once, instead of making admin to fix one error condition and then hit second one.

marccarre replied:
Resolved in latest version. Thanks for the feedback 👍

kad commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -316,6 +311,13 @@ func newInitData(cmd *cobra.Command, args []string, options *initOptions, out io
+		return nil, err
+	}
+	// Also set the union of pre-flight errors to InitConfiguration, to provide a consistent view of the runtime configuration:
+	cfg.NodeRegistration.IgnorePreflightErrors = ignorePreflightErrorsSet.List()
```
Also nit: do we have usage of this somewhere? Otherwise we are loosing prestine state as we read from config file, isn't it?

marccarre replied:
This is currently not used anywhere, but given the configuration object is passed around, and this field _could_ eventually be used elsewhere as a result, I thought doing the above might reduce the chance of a bug/inconsistent behaviour. I can remove it if you'd rather keep the configuration file's initial state. Let me know.

kad reviewed (commented):
small nits

rosti commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -463,20 +463,35 @@ func ValidateAPIEndpoint(c *kubeadm.APIEndpoint, fldPath *field.Path) field.Erro
+		// explicit list of errors they want to ignore, as it can be risky to
+		// mask all errors in such a way. Hence, we return an error:
+		allErrs = append(allErrs, field.Invalid(field.NewPath("ignorePreflightErrors"), "all", "'all' cannot be used in configuration file"))
+		return sets.NewString(), allErrs.ToAggregate()
```
As @kad mentioned below, let's drop this return and the one bellow and return `allErrs.ToAggregate()` in the last return at the end of the func.

marccarre replied:
Resolved in latest version. Thanks for the feedback 👍

rosti reviewed (commented):
Thanks @marccarre !
A couple of nits and we are there!

rosti commented:
Ok, the fuzzer test is failing, because we don't have the `IgnorePreflightErrors` in v1beta1. To fix this, we must reintroduce the `fuzzNodeRegistration` in the fuzzer test and plug the `IgnorePreflightErrors` field.
#75261 is the PR that got it deleted. You can check with it how to reintroduce it.
The new version of it should roughly look like this:
```golang
func fuzzNodeRegistration(obj *kubeadm.NodeRegistrationOptions, c fuzz.Continue) {	
	c.FuzzNoCustom(obj)	

 	// Pinning values for fields that get defaults if fuzz value is empty string or nil (thus making the round trip test fail)	
	obj.IgnorePreflightErrors = []string{}
}
```

Also, the reset test needs fixing (missing parameter in the call to `ValidateIgnorePreflightErrors`).

marccarre commented on cmd/kubeadm/app/cmd/join_test.go:
```diff
@@ -22,6 +22,7 @@ import (
 	"path/filepath"
 	"testing"
 
+	"k8s.io/apimachinery/pkg/util/sets"
```
It seems importing `"k8s.io/apimachinery/pkg/util/sets"` in tests yields to build errors like [that one](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/75499/pull-kubernetes-verify/1129715477703036929):
```diff
diff -Naupr -x _output /home/prow/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/BUILD /tmp/verify-bazel.STLhDZ/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/BUILD
--- /home/prow/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/BUILD	2019-05-18 11:50:23.049528531 +0000
+++ /tmp/verify-bazel.STLhDZ/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/BUILD	2019-05-18 11:51:00.467822797 +0000
@@ -99,6 +99,7 @@ go_test(
         "//staging/src/k8s.io/apimachinery/pkg/api/errors:go_default_library",
         "//staging/src/k8s.io/apimachinery/pkg/apis/meta/v1:go_default_library",
         "//staging/src/k8s.io/apimachinery/pkg/runtime:go_default_library",
+        "//staging/src/k8s.io/apimachinery/pkg/util/sets:go_default_library",
         "//staging/src/k8s.io/client-go/kubernetes/fake:go_default_library",
         "//staging/src/k8s.io/client-go/testing:go_default_library",
         "//staging/src/k8s.io/client-go/tools/clientcmd:go_default_library",

Run ./hack/update-bazel.sh
```
This presumably happens because Bazel (which I'm unfamiliar with) requires this dependency to be declared. I have added it manually in `cmd/kubeadm/app/cmd/BUILD` but is there a better way to achieve this?

rosti replied:
Yes, there is a script - `./hack/update-bazel.sh` that will do everything for you. Actually, always use that script to update bazel deps.

marccarre replied:
Thanks for the explanation! 👍

rosti reviewed (approved):
Thank you @marccarre !
I am happy how things turned out. Can you squash your commits as a final step so we can merge it?
/approve
/hold
For the squash.

marccarre commented:
Done: PR rebased on top of latest `master` and commits squashed.
Thanks again for the feedback! 👍

rosti commented:
@kad can you do a final pass and lgtm if you are happy with this?
/assign @kad 
/hold cancel
/test pull-kubernetes-e2e-gce

kad commented:
/lgtm
I have some doubts about some corner cases like in order of operations in upgrade steps (e.g. validating flags before/after trying to load cluster config), but I think it is not a big deal at the moment.
